### DISCUSSION TITLE
Can I switch from an in-memory to on-disk database?

### DISCUSSION THREAD

**Discussion by angsolcm** [sqlitebrowser/sqlitebrowser#3234] [Q&A] [answered]
Hello.

Let's say I've started off with a new in-memory database. After sinking some work into it I have decided I'd like to persist the database to disk. Is there a way to do this? I may be missing something, but there's nothing obvious I can see in the menus.

If not, it would be an appreciated addition!

Thanks.

**Comment by chrisjlocke** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/7872218/204530541-73a17d76-bc0f-4905-8772-3e3c4cb7a48e.png)

Click 'File' and 'Export' and 'Database to SQL file'.  This will backup your database to a text file.


![image](https://user-images.githubusercontent.com/7872218/204531021-5edd12d2-5d36-484f-830e-98978b78e37e.png)

Click 'File', 'Import' and 'Database from SQL file'.  You will be prompted to create a new file.

Just beware if you've got tables and relationships set up.  There is an 'issue' that if you have table A, and table B, and table B has foreign keys that rely on table A, you can't create table B before table A has been created, so you might have to juggle round the order of table creation.  I think there is also a pragma to ignore foreign key wobbliness too.

**Reply by angsolcm** [reply]
Thanks @chrisjlocke. You and @scottfurry have answered almost identically within the same minute of each other! I hope I can mark both as the answer.

**Reply by chrisjlocke** [reply]
Scotts had more bullet points. 😂

**Comment by scottfurry** [comment]
There's a bit of mouse-jujitsu required to make this happen. The challenge is that sqlite3 _assumes_ that the database is open/established.

After selecting `in-memory` from the menus and then deciding to save to a database file involves a few steps:

1. Start with a working `in-memory` database
2. Select File -> Export -> Database to SQL File
3. On the `Export SQL` dialog, click `Select All`, check all options, change drop boxes to `Export everything` and `Overwrite old schema`, then click `Save`
4. Save File Dialog pops up allowing you to save the SQL text to location of your choice.
5. Close the `in-memory` database.
6. Click `New Database` and click `cancel` on the create table dialog.
7. Go to the Execute SQL tab
8. Click `Open SQL` (second item from left in toolbar below Execute SQL tab title)
9. Navigate to Save SQL file from above.
10. Click `Execute` from SQL toolbar.
11. Select `Write Changes` on main toolbar.

You should now have the schema and data from an in-memory database saved as a file. A bit convoluted but this is the nature of sqlite.

**Reply by angsolcm** [reply]
Apologies @scottfurry . I can only mark 1 answer as _the_ answer.

**Comment by scottfurry** [comment]
@angsolcm - no worries. Some people need pictures, others need a blow-by-blow-do-this. So long as a positive result was achieved.

**Comment by mgrojo** [comment]
In the nightly version, you also have a "Save Database As..." option to do this directly. You have to commit your changes to memory ("Write Changes") for it to be enabled.

![imagen](https://github.com/sqlitebrowser/sqlitebrowser/assets/13812910/9b462872-3c11-43e8-90df-af79c4fd72ea)
